### ISSUE TITLE
Proposal: detect expensive `Logger` calls that should be refactored

### ISSUE THREAD

**Issue by jmitchell** [rrrene/credo#299]
### Environment

* Credo version: 0.5.3
* Erlang/Elixir version: Erlang/OTP 19; Elixir 1.4.0
* Operating system: Ubuntu 16.04.1 LTS

### What were you trying to do?

Detect any expensive calls to `Logger` that aren't wrapped in a zero-argument function (aka a thunk), and be advised to refactor.

### Expected outcome

This code

```elixir
Logger.debug "my_fun/1: input: #{inspect input}"
```

should be refactored to:

```elixir
Logger.debug fn ->
  "my_fun/1: input: #{inspect input}"
end
```

### Actual outcome

[This issue](https://github.com/jmitchell/backtrex/pull/1/files) wasn't detected during development.

**Comment by jmitchell** [comment]
For more context on what prompted this, see ["Logging: a silent performance killer"](https://elixirforum.com/t/logging-a-silent-performance-killer/3258).

**Comment by jmitchell** [comment]
After more reflection I realized there's room for improvement in Elixir itself. I submitted a [proposal](https://groups.google.com/d/msg/elixir-lang-core/LNZNyUGTdno/9Y0YOQ5tCQAJ) to elixir-lang-core. If accepted and implemented it may obviate the need for any work here.

**Comment by josevalim** [comment]
Given the Elixir behaviour, if accepted, would be opt-in, I believe this suggestion has enough merit on its own. :+1:

**Comment by rrrene** [comment]
@jmitchell I like this idea! If anybody wants to submit a PR on this, I would accept it :+1:

**Comment by rrrene** [comment]
This has been implemented in #326. Thx for the initial suggestion! :+1:
